stl/inc/xstring

```diff
@@ -4269,55 +4301,61 @@ template <class _Elem, class _Traits, class _Alloc>
 }
 
 template <class _Elem, class _Traits, class _Alloc>
 _NODISCARD basic_string<_Elem, _Traits, _Alloc> operator+(
     _In_z_ const _Elem* const _Left, const basic_string<_Elem, _Traits, _Alloc>& _Right) {
-    using _String_type = basic_string<_Elem, _Traits, _Alloc>;
-    using _Size_type   = typename _String_type::size_type;
-    _String_type _Ans;
-    _Ans.reserve(_Convert_size<_Size_type>(_Traits::length(_Left) + _Right.size()));
-    _Ans += _Left;
-    _Ans += _Right;
-    return _Ans;
+    using _Size_type       = typename basic_string<_Elem, _Traits, _Alloc>::size_type;
+    const auto _Left_size  = _Convert_size<_Size_type>(_Traits::length(_Left));
+    const auto _Right_size = _Right.size();
+    if (_Right.max_size() - _Right_size < _Left_size) {
+        _Xlen_string();
+    }
+
+    return {_String_constructor_concat_tag{}, _Right, _Left, _Left_size, _Right.c_str(), _Right_size};
 }
 
 template <class _Elem, class _Traits, class _Alloc>
 _NODISCARD basic_string<_Elem, _Traits, _Alloc> operator+(
     const _Elem _Left, const basic_string<_Elem, _Traits, _Alloc>& _Right) {
-    basic_string<_Elem, _Traits, _Alloc> _Ans;
-    _Ans.reserve(1 + _Right.size());
-    _Ans += _Left;
-    _Ans += _Right;
-    return _Ans;
+    const auto _Right_size = _Right.size();
+    if (_Right_size == _Right.max_size()) {
+        _Xlen_string();
+    }
+
+    return {_String_constructor_concat_tag{}, _Right, _STD addressof(_Left), 1, _Right.c_str(), _Right_size};
 }
 
 template <class _Elem, class _Traits, class _Alloc>
 _NODISCARD basic_string<_Elem, _Traits, _Alloc> operator+(
     const basic_string<_Elem, _Traits, _Alloc>& _Left, _In_z_ const _Elem* const _Right) {
-    using _String_type = basic_string<_Elem, _Traits, _Alloc>;
-    using _Size_type   = typename _String_type::size_type;
-    _String_type _Ans;
-    _Ans.reserve(_Convert_size<_Size_type>(_Left.size() + _Traits::length(_Right)));
-    _Ans += _Left;
-    _Ans += _Right;
-    return _Ans;
+    using _Size_type       = typename basic_string<_Elem, _Traits, _Alloc>::size_type;
+    const auto _Left_size  = _Left.size();
+    const auto _Right_size = _Convert_size<_Size_type>(_Traits::length(_Right));
+    if (_Left.max_size() - _Left_size < _Right_size) {
+        _Xlen_string();
+    }
+
+    return {_String_constructor_concat_tag{}, _Left, _Left.c_str(), _Left_size, _Right, _Right_size};
 }
 
 template <class _Elem, class _Traits, class _Alloc>
 _NODISCARD basic_string<_Elem, _Traits, _Alloc> operator+(
     const basic_string<_Elem, _Traits, _Alloc>& _Left, const _Elem _Right) {
-    basic_string<_Elem, _Traits, _Alloc> _Ans;
-    _Ans.reserve(_Left.size() + 1);
-    _Ans += _Left;
-    _Ans += _Right;
-    return _Ans;
+    const auto _Left_size = _Left.size();
+    if (_Left_size == _Left.max_size()) {
+        _Xlen_string();
+    }
+
+    return {_String_constructor_concat_tag{}, _Left, _Left.c_str(), _Left_size, _STD addressof(_Right), 1};
 }

```

Comment:
Pre-existing: I observe that `basic_string&& + basic_string&&` below is still calling `.size()` multiple times instead of factoring it out.

Reply:
Wow, and also it never tries to use _Right's buffer if that would be cheaper. And also it doesn't always comply with P1165R1.

Reply:
Done.